GSoC2022 - Interpolated Corrected Curvatures (#6760)
CGAL/cgal: Polygon_mesh_processing/doc/Polygon_mesh_processing/Polygon_mesh_processing.txt

\anchor Chapter_PolygonMeshProcessing

\cgalAutoToc

> \authors David Coeurjolly, Jaques-Olivier Lachaud,  Konstantinos Katriopla, Sébastien Loriot, Mael Rouxel-Labbé, Hossam Saeed, Jane Tournois, and Ilker %O. Yaz

Reviewer: ```suggestion
\authors David Coeurjolly, Jaques-Olivier Lachaud,  Konstantinos Katrioplas, Sébastien Loriot, Mael Rouxel-Labbé, Hossam Saeed, Jane Tournois, and Ilker %O. Yaz
```

\cgalExample{Polygon_mesh_processing/compute_normals_example_Polyhedron.cpp}

****************************************
\section PMPICC Computing Curvatures

This package provides methods to compute curvatures on polygonal meshes based on Interpolated Corrected Curvatures

> on Polyhedral Surfaces \cgalCite{lachaud2020}. This includes mean curvature, gaussian curvature,

Author: Should It be applied to variable & function names as well?
Or just wherever gaussian curvature is mentioned in comments or documentation?

> on Polyhedral Surfaces \cgalCite{lachaud2020}. This includes mean curvature, gaussian curvature,

Reviewer 2: Basically everything that is public (doc or function name)

- `CGAL::Polygon_mesh_processing::interpolated_corrected_principal_curvatures_and_directions()`
- `CGAL::Polygon_mesh_processing::interpolated_corrected_curvatures()`

Where it is recommended to use the last function for computing multiple curvatures (example: mean and gaussian)
as the implementation performs the shared computations only once, making it more efficient.

Similarly, we can use the following functions to compute curvatures on a specific vertex:
- `CGAL::Polygon_mesh_processing::interpolated_corrected_mean_curvature_one_vertex()`
- `CGAL::Polygon_mesh_processing::interpolated_corrected_Gaussian_curvature_one_vertex()`
- `CGAL::Polygon_mesh_processing::interpolated_corrected_principal_curvatures_and_directions_one_vertex()`
- `CGAL::Polygon_mesh_processing::interpolated_corrected_curvatures_one_vertex()`

\cgalFigureRef{icc_diff_radius} shows how the mean curvature changes depending on
the ball_radius named parameter which can be set to a value > 0 to get a smoother
distribution of values and "diffuse" the extreme values of curvatures across the mesh.

\cgalFigureAnchor{icc_diff_radius}
<center>

> <img src="icc_diff_radius.png" style="max-width:90%;"/>

Author: I edited the image and made the captions smaller, and kept the captions in as I thought it makes them easier to read. I think it looks better now. If you want to move the text out of the image to the text caption, please let me know.

> <img src="icc_diff_radius.png" style="max-width:90%;"/>

Reviewer 2: I'm wondering if it makes senses to have caption in the figure and not as plain text in the doxygen caption section of the figure.

</center>
\cgalFigureCaptionBegin{icc_diff_radius}
The mean curvature distribution on a bear mesh with different values for the expanding ball radius
\cgalFigureCaptionEnd

Property maps are used to record the computed curvatures as shown in examples.

> Property maps are an API introduced in the boost library that allows associating

Reviewer 2: not yet addressed

\cgalExample{Polygon_mesh_processing/compute_normals_example_Polyhedron.cpp}

****************************************
\section PMPICC Computing Curvatures

This package provides methods to compute curvatures on polygonal meshes based on Interpolated
Corrected Curvatures on Polyhedral Surfaces \cgalCite{cgal:lrtc-iccmps-20}. This includes mean curvature,
Gaussian curvature, principal curvatures and directions. These can be computed on triangle meshes,
quad meshes, and meshes with n-gon faces (for n-gons, the centroid must be inside the n-gon face).
The algorithms used prove to work well in general. Also, on meshes with noise on vertex positions,

> they give accurate results, on the condition that the correct vertex normals are provided.

Reviewer: ```suggestion
they give accurate results, under the condition that the correct vertex normals are provided.
```

\subsection ICCBackground Brief Background

Surface curvatures are quantities that describe the local geometry of a surface. They are important in many
geometry processing applications. As surfaces are 2-dimensional objects (embedded in 3D), they can bend
in 2 independent directions. These directions are called principal directions, and the amount of bending
in each direction is called the principal curvature: \f$ k_1 \f$ and \f$ k_2 \f$ (denoting max and min
curvatures). Curvature is usually expressed as scalar quantities like the mean curvature \f$ H \f$ and
the Gaussian curvature \f$ K \f$ which are defined in terms of the principal curvatures.

The algorithms are based on the two papers \cgalCite{cgal:lrt-ccm-22} and \cgalCite{cgal:lrtc-iccmps-20}. They
introduce a new way to compute curvatures on polygonal meshes. The main idea in \cgalCite{cgal:lrt-ccm-22} is
based on decoupling the normal information from the position information, which is useful for dealing with
digital surfaces, or meshes with noise on vertex positions. \cgalCite{cgal:lrtc-iccmps-20} introduces some

> extensions to this framework. As it uses linear interpolation on the corrected normal vector field

Reviewer: ```suggestion
extensions to this framework, as it uses linear interpolation on the corrected normal vector field
```

  \mu^{\mathbf{X},\mathbf{Y}}(\tau_{ijk}) = & \frac{1}{2} \big\langle \bar{\mathbf{u}} \big| \langle \mathbf{Y} | \mathbf{u}_k -\mathbf{u}_i \rangle \mathbf{X} \times (\mathbf{x}_j - \mathbf{x}_i) \big\rangle
    -\frac{1}{2} \big\langle \bar{\mathbf{u}} \big| \langle \mathbf{Y} | \mathbf{u}_j -\mathbf{u}_i \rangle \mathbf{X} \times (\mathbf{x}_k - \mathbf{x}_i) \big\rangle,
  \end{align*}
\f]
where \f$ \langle \cdot \mid \cdot \rangle \f$ denotes the usual scalar product,
\f$ \bar{\mathbf{u}}=\frac{1}{3}( \mathbf{u}_i + \mathbf{u}_j + \mathbf{u}_k )\f$.

The first measure \f$ \mu^{(0)} \f$ is the area measure of the triangle, and the measures \f$ \mu^{(1)} \f$ and
\f$ \mu^{(2)} \f$ are the mean and Gaussian corrected curvature measures of the triangle. The last measure
\f$ \mu^{\mathbf{X},\mathbf{Y}} \f$ is the anisotropic corrected curvature measure of the triangle. The
anisotropic measure is later used to compute the principal curvatures and directions through an eigenvalue
solver.

The interpolated curvature measures are then computed for each vertex \f$ v \f$ as the sum of
the curvature measures of the faces in a ball around \f$ v \f$ weighted by the inclusion ratio of the

> triangle in the ball. if the ball radius is not specified, the sum is instead over the incident faces

Reviewer: ```suggestion
triangle in the ball. If the ball radius is not specified, the sum is instead computed over the incident faces
```

of \f$ v \f$.

To get the final curvature value for a vertex \f$ v \f$, the respective interpolated curvature measure
is divided by the interpolated area measure.

\f[
\mu^{(k)}( B ) = \sum_{\tau : \text{triangle} } \mu^{(k)}( \tau ) \frac{\mathrm{Area}( \tau \cap B )}{\mathrm{Area}(\tau)}.
\f]

\subsection ICCAPI API

The implementation is generic in terms of mesh data structure. It can be used on `Surface_mesh`,
`Polyhedron_3` and other polygonal mesh structures based on the concept `FaceGraph`.

These computations are performed using (on all vertices of the mesh) `CGAL::Polygon_mesh_processing::interpolated_corrected_curvatures()`

> where function named parameters are used to select the curvatures (and possibly directions) to be computed. An overlay with the same name

Reviewer 3: overlay or overload ??

> where function named parameters are use to select the curvatures (and possibly directions) to be computed. An overlay with the same named
> but accepting and accepting a given vertex is also available in case the computation should be done only for that vertex.

Reviewer: ```suggestion
where function named parameters are used to select the curvatures (and possibly directions) to be computed. An overlay with the same name
but taking a given vertex is also available in case the computation should be done only for that vertex.
```

\subsection ICCResults Results & Performance

> **To be updated**

Reviewer: Update the results or remove this line?

Author: Should be removed. I will do that shortly

Reviewer 2: Moved to #7859

> First, \cgalFigureRef{icc_measures} first illustrates various curvature measure on a triangular mesh.

Reviewer: ```suggestion
First, \cgalFigureRef{icc_measures} illustrates various curvature measures on a triangular mesh.
```

\cgalFigureAnchor{icc_measures}
<center>
<table border=0>
<tr>
<td><img src="bimba-mean0.040000-0.000000.png" style="width:100%;"/></td>
<td><img src="bimba-gaussian0.040000-0.000000.png" style="width:100%;"/></td>
<td><img src="bimba-dmin0.040000-0.000000.png" style="width:100%;"/></td>
<td><img src="bimba-dmax0.040000-0.000000.png" style="width:100%;"/></td>
</tr>
<tr align="center"><td>(a)</td><td>(b)</td></tr>
</table>
</center>

\cgalFigureCaptionBegin{icc_measures}

> Mean curvature, Gaussian curvature, minimal principal curvature direction and minimal principal curvature direction on a mesh (ball radius set to 0.04).

Reviewer: ```suggestion
Mean curvature, Gaussian curvature, minimal principal curvature direction and maximal principal curvature direction on a mesh (ball radius set to 0.04).
```

\cgalFigureCaptionEnd

\ref BGLPropertyMaps are used to record the computed curvatures as shown in examples. In the following examples, for each property map, we associate
a curvature value to each vertex.

\subsection ICCExampleSM Interpolated Corrected Curvatures on a Surface Mesh Example

The following example illustrates how to
compute the curvatures on vertices
and store them in the property maps provided by the class `Surface_mesh`.

\cgalExample{Polygon_mesh_processing/interpolated_corrected_curvatures_SM.cpp}

\subsection ICCExamplePH Interpolated Corrected Curvatures on a Polyhedron Example

The following example illustrates how to
compute the curvatures on vertices
and store them in dynamic property maps as the class `Polyhedron_3` does
not provide storage for the curvatures.

\cgalExample{Polygon_mesh_processing/interpolated_corrected_curvatures_PH.cpp}

\subsection ICCExampleSV Interpolated Corrected Curvatures on a Vertex Example

The following example illustrates how to

> compute the curvatures on a specific vertex

Reviewer: ```suggestion
compute the curvatures on a specific vertex.
```
